core/environment/fsm_test.go

```diff
@@ -0,0 +1,288 @@
+				env.Sm.SetState("STANDBY")
+				Expect(env.Sm.Can("DEPLOY")).To(BeTrue())
+				Expect(env.Sm.Can("GO_ERROR")).To(BeTrue())
+				Expect(env.Sm.Can("EXIT")).To(BeTrue())
+			})
+			It("should not allow for other transitions", func() {
+				env.Sm.SetState("STANDBY")
+				Expect(env.Sm.Cannot("CONFIGURE")).To(BeTrue())
+				Expect(env.Sm.Cannot("RESET")).To(BeTrue())
+				Expect(env.Sm.Cannot("START_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("STOP_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RECOVER")).To(BeTrue())
+			})
+		})
+		When("FSM is in DEPLOYED", func() {
+			It("should allow for CONFIGURED, GO_ERROR and EXIT transitions", func() {
+				env.Sm.SetState("DEPLOYED")
+				Expect(env.Sm.Can("CONFIGURE")).To(BeTrue())
+				Expect(env.Sm.Can("GO_ERROR")).To(BeTrue())
+				Expect(env.Sm.Can("EXIT")).To(BeTrue())
+			})
+			It("should not allow for other transitions", func() {
+				env.Sm.SetState("DEPLOYED")
+				Expect(env.Sm.Cannot("DEPLOY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RESET")).To(BeTrue())
+				Expect(env.Sm.Cannot("START_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("STOP_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RECOVER")).To(BeTrue())
+			})
+		})
+		When("FSM is in CONFIGURED", func() {
+			It("should allow for START_ACTIVITY, RESET, GO_ERROR and EXIT transitions", func() {
+				env.Sm.SetState("CONFIGURED")
+				Expect(env.Sm.Can("START_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Can("RESET")).To(BeTrue())
+				Expect(env.Sm.Can("GO_ERROR")).To(BeTrue())
+				Expect(env.Sm.Can("EXIT")).To(BeTrue())
+			})
+			It("should not allow for other transitions", func() {
+				env.Sm.SetState("CONFIGURED")
+				Expect(env.Sm.Cannot("DEPLOY")).To(BeTrue())
+				Expect(env.Sm.Cannot("CONFIGURE")).To(BeTrue())
+				Expect(env.Sm.Cannot("STOP_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RECOVER")).To(BeTrue())
+			})
+		})
+		When("FSM is in RUNNING", func() {
+			It("should allow for STOP_ACTIVITY and GO_ERROR transitions", func() {
+				env.Sm.SetState("RUNNING")
+				Expect(env.Sm.Can("STOP_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Can("GO_ERROR")).To(BeTrue())
+			})
+			It("should not allow for other transitions", func() {
+				env.Sm.SetState("RUNNING")
+				Expect(env.Sm.Cannot("DEPLOY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RESET")).To(BeTrue())
+				Expect(env.Sm.Cannot("CONFIGURE")).To(BeTrue())
+				Expect(env.Sm.Cannot("START_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RECOVER")).To(BeTrue())
+				Expect(env.Sm.Cannot("EXIT")).To(BeTrue())
+			})
+		})
+		When("FSM is in ERROR", func() {
+			It("should allow for RECOVER transition", func() {
+				env.Sm.SetState("ERROR")
+				Expect(env.Sm.Can("RECOVER")).To(BeTrue())
+				// fixme: is this correct that we do not allow for EXIT?
+				//  in principle, in TeardownEnvironment we ignore the FSM and do it ourselves,
+				//  but I don't see a reason not to use allow for it
```

Comment:
An arbitrary choice I think. Currently we have no ERROR recovery, and this FSM maps to task states. But killing the task is possible even when that task's FSM disobeys or doesn't work.